### ISSUE TITLE
Many-to-many (non-unique) relationships.

### ISSUE THREAD

**Issue by marcreicher** [sequelize/sequelize#5077]
Hey there. I am working on creating a trivia-style website (and I am using Sequelize and MySQL). I have three data tables. A table of Users, a table of Quizzes, and a junction table (UserQuizzes). I was hoping for users to be able to take the same quiz multiple times and have all of their scores to the UserQuizzes table. Thus, I need the junction table to allow for duplicate posts of User/Quiz pairs. When defining my model relationships, I had used the following lines of code.

```
  User.belongsToMany(Quiz, {through: 'UserQuiz'});
  Quiz.belongsToMany(User, {through: 'UserQuiz'});
```

This resulted in a UserQuiz table that did not allow for duplicate User/Quiz pairs (as the primary key of the UserQuiz table was the UserID/QuizID combination). I did some research, and I read somewhere that Sequelize does not support non-unique NxM associations (not sure if that is still the case though, so please correct me if I am wrong). Regardless, I worked around this issue by 'manually' creating the UserQuiz table like so:

```
sequelize.define('UserQuiz', {
    id: {
      type: DataTypes.INTEGER(20),
      primaryKey: true,
      autoIncrement: true
    },
    userID: {
      type: DataTypes.INTEGER(11).UNSIGNED,
      references: {
        model: 'User',
        key: 'id'
      },
      field: 'user_id',
      allowNull: false
    },
    quizID: {
      type: DataTypes.INTEGER(11).UNSIGNED,
      references: {
        model: 'Quiz',
        key: 'id'
      },
      field: 'quiz',
      allowNull: false
    },
    timeToComplete: {
      type: DataTypes.DECIMAL(13, 2).UNSIGNED,
      field: 'time_to_complete'
    }
```

This worked exactly as I had hoped. However, I am now running into some issues when trying to query the UserQuiz table. My goal is to query the table and find the five fastest times for any given Quiz. So, I used a query like this:

```
UserQuiz.findAll({
      where: {
        challengeID: 2
      },
      limit: 5,
      order: 'timeToComplete ASC'
    })
```

This also worked. However, I was also interested in pulling information from the User table as well. I read through the Sequelize docs about 'eager loading' and tried out a query like this:

```
UserQuiz.findAll({
      where: {
        challengeID: 2
      },
      limit: 5,
      order: 'timeToComplete ASC'
      include: [{
            model: User,
            as: 'userID'
      }]
    })
```

I am not able to get the 'include' statement to ever work. Could this be happening because I am no longer defining model relationships with 'belongsTo' statements? Regardless, is there any way to use NxM non-unique associations and still take advantage of eager loading? Or am I trying to pull off something that Sequelize does not currently support?

**Comment by mickhansen** [comment]
non-unique N:M are currently poorly supported in Sequelize. The feature lacks adequate tests to cover most cases, and obviously code for those cases.

Your best bet is to handle the join table manually.

**Comment by xxxtonixxx** [comment]
@mickhansen is non-unique N:M supported already? I have same problem...

**Comment by mickhansen** [comment]
@xxxtonixxx Not completely, per the issue being open, it's not really a significant issue in terms of volume and has adequate work arounds.

**Comment by felixfbecker** [comment]
I'm pretty sure though that the approach @marcreicher took initially works. He just used the `as` option wrong.

**Comment by xxxtonixxx** [comment]
@mickhansen and @felixfbecker , then how can I avoid unique key in my table? I have two models that generate a third model (N:M model), but my mysql table is generated with this `UNIQUE KEY country_language_id_language_id_country_unique (id_country,id_language)`. I have same relations than @marcreicher , using `belongsToMany()`
```javascript
    Language.belongsToMany(Country, {
      through: CountryLanguage,
      foreignKey: {
        name: 'id_language',
        allowNull: false
      }
    });

    Country.belongsToMany(Language, {
      through: CountryLanguage,
      foreignKey: {
        name: 'id_country',
        allowNull: false
      }
    });
```
Do I need to use `references` attribute in my model instead of `belongsToMany`? If it's so, I lose any functionality? I suppose I lose `getLanguages()`, `getCountries()`, etc. methods.

Thanks for your fast reply :smile:

**Comment by mx781** [comment]
@mickhansen could you (or anyone reading) elaborate on the different workarounds for this? 

While I've looked through various related issues (https://github.com/sequelize/sequelize/issues/5193, https://github.com/sequelize/sequelize/issues/3220, https://github.com/sequelize/sequelize/issues/6906) that touch on non-unique N:M relationships, neither approach really works if I am querying a different model from the "through model"(CountryLanguage in post above, or UserQuiz in OP's example). 

I thought of using a `beforeFindAfterOptions` hook, but I can't seem to find an option that would prevent the de-duplication.

Is there a way for something like [this](https://gist.github.com/mx781/dc6b86fad71c394872834397c3f8a52f) to work? The goal is to retrieve a User and their Bookings without de-duping the Resources for each Booking.

**Comment by xxxtonixxx** [comment]
Not resolved

**Comment by ilikebits** [comment]
This is still broken. We had to work around this in our codebase by manually `include`ing the join table, and then doing a nested `include` on the target table.

**Comment by Kxze** [comment]
Still an issue.

**Comment by kunyan** [comment]
Anyone are focus on this feature ?

**Comment by smddzcy** [comment]
It's still an issue. Waiting for a proper solution.

**Comment by joeythomaschaske** [comment]
yeeeshhh still open 4 years later?

**Comment by MohammadHasham** [comment]
Still open...

**Comment by sgarcia30** [comment]
I guess it's obviously still an issue.  I'm having trouble with my query returning duplicates when I make a query that includes a `hasMany` join...

**Comment by sprietNathanael** [comment]
Well, still an issue apparently. Is there a workaround for this?
I have two tables `A` and `B` that can be linked multiple times at different periods. I defined `AB` table with an `id` as primary key (`primarykey`, `unique`, `autoincrement`) and with `dateFrom` and `dateTo` fields.
Is there a way to add `dateFrom` and `dateTo` to the `unique` constraint between `Aid` and `Bid`? Or simply avoid this constraint?

**Comment by RodrigoLeiteF** [comment]
My solution was to stop using ORMs entirely. I find that they too often get in the way instead of helping, so I decided to use a regular SQL driver and map the responses to Typescript interfaces / types myself.

**Comment by papb** [comment]
Doesn't creating a new separated for this works around the problem?

**Comment by tresloukadu** [comment]
> My solution was to stop using ORMs entirely. I find that they too often get in the way instead of helping, so I decided to use a regular SQL driver and map the responses to Typescript interfaces / types myself.

Still the same problem. There is a lot of suffering to join two tables through a third one. You gave a good point to think about : "stop using ORM", I will reflect about it because sequelize is slowing down my developing time.

**Comment by codecamp-official-thailand** [comment]
https://stackoverflow.com/a/58439584

**Comment by abhi5658** [comment]
>@mx781:
> @mickhansen could you (or anyone reading) elaborate on the different workarounds for this?
> 
> While I've looked through various related issues (#5193, #3220, #6906) that touch on non-unique N:M relationships, neither approach really works if I am querying a different model from the "through model"(CountryLanguage in post above, or UserQuiz in OP's example).
> 
> I thought of using a `beforeFindAfterOptions` hook, but I can't seem to find an option that would prevent the de-duplication.
> 
> Is there a way for something like [this](https://gist.github.com/mx781/dc6b86fad71c394872834397c3f8a52f) to work? The goal is to retrieve a User and their Bookings without de-duping the Resources for each Booking.

Refactored the code to not have de-duplication
Changes in the @mx781  gist's comment https://gist.github.com/mx781/dc6b86fad71c394872834397c3f8a52f#gistcomment-3748566

**Comment by LeonxLJX** [comment]
I'd like to take this one.

Reading the issue, this looks like a documentation-clarity gap rather than a code defect: the current page doesn't give readers enough to self-diagnose.

Plan: locate the relevant section, confirm what the code actually does today (so the docs match behavior, not intent), tighten the wording, and scan neighboring pages for the same inconsistency — otherwise the fix just relocates the confusion.

One thing to confirm before writing: minimal correction scoped to this page, or also cross-link from related guides if the same gap shows up there?

(claiming via @LeonxLJX)

**Comment by DeployIsComing** [comment]
Thanks I have removed sequelize from my project and moved to knex

please close this thread
